[USB] Power up immediately when booting with USB connected

fiam opened the pull request:
This allows jumping out of ST's bootloader without manual user
intervention when the radio is connected to the USB host while
it's powered off.

Note that the radio doesn't power on as soon as it's connected, since
the hardware is wired to boot the MCU into bootloader mode. However,
the host can detect the USB connection and issue a DFU_DNLOAD to
turn on the radio.

fiam committed:
[USB] Power up immediately when booting with USB connected

This allows jumping out of ST's bootloader without manual user
intervention when the radio is connected to the USB host while
it's powered off.

Note that the radio doesn't power on as soon as it's connected, since
the hardware is wired to boot the MCU into bootloader mode. However,
the host can detect the USB connection and issue a DFU_DNLOAD to
turn on the radio.

kilrah commented:
Again there is a lack of description of the functionality you intend to achieve with this. 
We have to guess, and from our guess this implementation would not be optimal...

fiam commented:
@kilrah You've made OpenTX's position clear. I'm closing these PRs in the meantime. No reason to waste my time arguing.

kilrah commented:
When a random guy comes and wants to change stuff with no apparent reason it's kinda normal to request information don't you think?

Present yourself, make your intentions clear, explain in which way you'd want to contribute to improving the project and we'll be able to work together. 

@3djc actually spent a day working to rework one of your PRs so that it fits better with the code style and structure to show you how, and you closed some that were even already accepted.

fiam commented:
>@3djc actually spent a day working to rework one of your PRs so that it fits better with the code
> style and structure to show you how.

You can't say that with a straight face. @3djc took the code from #7202, fixed the compilation error on MSVC (which I could have fixed myself, but you guys instead closed the PR), removed the comments and then mixed it with something entirely different that shouldn't even go into the same PR, much less into the same commit (unifying the horus and taranis drivers doesn't have anything to do with fixing ST-LINK usage on the X7). Then submitted it as #7212 as if he had done it all himself.

What's to show there? Outsiders not welcome?

bsongis commented:
If I am not mistaken 3djc took the #7202 and improved it, he reused the code from Horus instead of duplicating it. It's another PR, feel free to comment, it's public. It won't be merged before next monday as we plan to release 2.3.3 at the end of the week.

bsongis commented:
Also some of your PRs were accepted directly ... some were accepted after you reworked them... That's the game. Outsiders are very welcome, as you can see in the log history!

fiam commented:
@bsongis I've never expected my PRs to be accepted as is. In fact, I was very patient in #7186 explaining the reasoning why I chose the solution I had initially submitted, asked for clarifications without getting any and amended the code after it became clear to me that there was no reason to use the host compiler.

I didn't even complain when you closed #7202 saying that you would take over it. I understand sometimes redoing something yourself in a different way is easier and faster than explaining the changes you feel that should be made.

However, that's not what happened in #7202. All code but the comments ended up in #7212, just fixing the trivial error when compiling with MSVC. Then, the PWR drivers for horus and taranis were merged into a single file, which is logically a completely different change and should have been done in a separate commit and even a separate PR. And, of course, there was no real reason to strip the comments away, that's just bad practice.

Saying that it was to "show me how" is naïve at best and dishonest at worst. And you know what's worse? It took me 5 minutes to find #6889, which was closed without merging but its code made its way into OpenTX in 746a2ae528 and 85a2b138f7 again without any mentions of the original author. So this is definitely a single occurrence of an unfortunate oversight.

It's perfectly understandable that sometimes you might want to make changes to someone else's PR before it gets merged into the repository. But the proper way to do that is to pull from their branch and add your commits on top. That preserves the credits of the original author. And you know what? Looks like someone is perfectly aware of that workflow, because that's how it was done in #7128.

I think we can all agree that intentionally passing code from others as yours is despicable. Open source doesn't mean you're allowed to take credit from others for you. And every open source project should make its best to ensure that the author of a piece of code is properly credited. Otherwise, what moral authority do you have if someone else takes the project and creates a closed source derivative? None, because you've been already taking credit from others.

Now regarding this PR and #7204, all I have to say is that asking me to provide the host side functionality at the same time as the supporting code in the radio side is a reason totally made up on the spot. There are tons of features that exist on OpenTX to interact with 3rd party hardware and software, but are entirely ignored or unsupported by the companion: HID, bluetooth, S.Port mirror, etc... just to name a few that are already there. Even the upcoming #6996 contains support for allowing a USB host to send channel values to the radio and I don't see anyone asking for the host side to be shown or added to companion before that functionality can be added to OpenTX.

Of course, I do understand that this is your project and these are your rules. I respect and accept that, no hard feelings from me. I'm just saying that they look very unfair and arbitrary, and that's why I'm choosing not to play by them.

bsongis commented:
I am trying to understand what you are saying. You seem full time on opentx while I am not, I am very sorry if you feel I am slow at answering to your questions. I also feel that some questions from Kilrah haven't been answered, but I can be wrong. As I said I am not full time on opentx.

I asked to the author of #7212 to add the comment you are referring to (I modified it a little bit though). This PR is still open, I see a reference to #7202. There is no intention to hide your contribution, your other PRs wouldn't have been merged otherwise. It happens that I rewrite everything when I see something good, but which could be improved. No bad intention, at least by me, and I doubt 3djc was any different here.

About #6889 I will need more context, did you find the fix yourself? Who is the original author? 

I don't think we are allowed to work on your repo. We are rather used to open an access for the regular contributors so that they can have Travis feedback before opening a PR, and then we can also add commits on the top. We are used to give credits and thank all people who have helped the project, please have a look to the release not if you feel I am wrong.

About #7204 you said it's a feature in progress. I don't understand why we would accept something which is not finished and that we can't review. Today it seems you are just introducing a Trojan horse into OpenTX, but without any explanation, that's just guess, again.

raphaelcoeffic commented:
> Of course, I do understand that this is your project and these are your rules. I respect and accept that, no hard feelings from me. I'm just saying that they look very unfair and arbitrary, and that's why I'm choosing not to play by them.

@fiam let's work together to get something that works for everyone, please!

Just for you to know: there is something inherent to OpenTX, which is it's embracing of open source. So, when you come up with some changes that target compatibility with some close source application you want to come up with later, for some obscure reasons (no explanations so far), then well, yes, the reactions are going to be harsch, sorry for that.

However, if you choose to open source your "firmware updater" (which I think this is going to be), and we can see it together with the changes in the PR, then I can bet with you that we will find a solution that is good for everybody.

On the other side, if the only application of some code change is to support a closed source application that does not even exist yet, that might check TX serial numbers (for whatever reason), then basically "only over my dead cold body" :-D

rotorman commented:
I stumbled upon this PR and find the idea interesting of automatically leaving DFU with a brief (5 to 10 second) timeout.

Allow me to explain where it would be useful. @olliw42 has implemented in his latest OpenTX [mavtelem fork](https://github.com/olliw42/opentx/tree/mavtelem2311-v24) a MAVLink router (video explaining it [here](https://youtu.be/s8XACNDZF_Y?t=107)) that allows the T16/TX16S top USB-C port to be used as virtual serial port for connecting up to GCS (e.g. Mission Planner, QGC).

I created a diagram that explains the usage scenario:
![grafik](https://raw.githubusercontent.com/rotorman/opentx/mavtelem-dev/doc/OlliW-OpenTX-TX16S-DL-GCS_USB.jpg)

**With GCS connected via USB-C top port, it would IMO be beneficial if the DFU mode is left after some seconds (e.g. 5 to 10) in order not to have to disconnect and re-connect the cable every time in a setup where GCS is permanently attached to radio (consider tray setup, with GCS PC/tablet on FPV mount as in the picture below).**
![grafik](https://user-images.githubusercontent.com/21011587/107872914-42dbd480-6eae-11eb-8ed5-14b04823ed3f.png)

Even though MAVLink router is not included in the main OpenTX branch, and thus this being for vanilla OpenTX users presently not relevant, it might be an option to consider, of not having to show a blank screen for user indefinitely if he/she forgets USB cable connected at power-up.

With STM32 chips, this must be possible, as e.g. ArduPilot running on similar STM32F4 controllers first goes automatically to bootloader, and then after a timeout to normal USB CDC/VCP mode. DFU is reached by holding a button on power-up.

Buit it might be something that the schematic of OpenTX radios is doing differently though. I had a look at [T16 schematics](https://github.com/opentx/opentx/files/5035629/T16.Schematic.pdf) and see that BOOT0 is directly connected via voltage divider (2 resistors) to USB VBUS, thus gets pulled up, whenever the USB is attached, meaning DFU is selected. And to leave DFU in this hard-wired BOOT0 case, I understand, only a USB host can trigger according to [STM32 docu here](https://www.st.com/resource/en/application_note/cd00264379-usb-dfu-protocol-used-in-the-stm32-bootloader-stmicroelectronics.pdf).
